### ISSUE TITLE
Torch + LMDB, not enough memory

### ISSUE THREAD

**Issue by DmitryUlyanov** [NVIDIA/DIGITS#659]
Hello, I am trying to use lmdb dataset that I created with DIGITS and torch defined model. The dataset has around 300k objects. The loading is interrupted by the error: 

```
2016-03-26 09:30:27 [INFO ] creating data readers
2016-03-26 09:30:27 [INFO ] opening LMDB database: /place/home/dmitry/git/DIGITS/digits/jobs/20160325-003220-9caf/train_db
2016-03-26 09:30:27 [INFO ] opening LMDB database: /place/home/dmitry/git/DIGITS/digits/jobs/20160325-003220-9caf/train_db
2016-03-26 09:30:27 [INFO ] opening LMDB database: /place/home/dmitry/git/DIGITS/digits/jobs/20160325-003220-9caf/train_db
2016-03-26 09:30:27 [INFO ] opening LMDB database: /place/home/dmitry/git/DIGITS/digits/jobs/20160325-003220-9caf/train_db
2016-03-26 09:32:08 [FAIL] /home/dmitry/torch/install/share/lua/5.1/threads/threads.lua:264:
[thread 4 callback] /place/home/dmitry/git/DIGITS/tools/torch/data.lua:326: not enough memory
```

I have tried to install different versions of LMDB lib, but it does not help. Thank you in advance.

**Comment by gheinrich** [comment]
Thanks for the report @DmitryUlyanov. And sorry to hear that you met this issue! I will try and reduce the memory footprint of DIGITS's data loader for Torch. In the meantime you might want to check whether reducing the number of loader threads helps with the memory issue. The default is 4 threads and this is coming from this [line](https://github.com/NVIDIA/DIGITS/blob/master/tools/torch/main.lua#L221) (for the training set) and this [line](https://github.com/NVIDIA/DIGITS/blob/master/tools/torch/main.lua#L235) (for the validation set).

**Comment by gheinrich** [comment]
Hi @DmitryUlyanov I notice that right after opening LMDB files, each thread allocates large amounts of memory in order to store keys to all entries. This takes in the order of 500MB worth of memory per thread per million keys. Did it help to reduce the number of threads? How much spare memory do you have on your system?

**Comment by DmitryUlyanov** [comment]
Hello, thank you for the investigation. I did not have time to try to reduce the number of workers yet. I have 12GB GPU, AlexNet with 128 batch, which I assume needs ~4 GB. The server has 256GB RAM, so problem is not there probably. 

I tried to run MNIST example and it worked good for me. I will try reduce the number of workers in a while and report. Thanks!

**Comment by TimZaman** [comment]
Just to bump this up, i added a variable to https://github.com/NVIDIA/DIGITS/blob/master/tools/torch/main.lua#L221 and https://github.com/NVIDIA/DIGITS/blob/master/tools/torch/main.lua#L235 so that the amount of used threads are now **1** instead of 4. I am running this with a Titan X; and it is using **5GB** for ImageNet, 1.3 million images of 256x256. LMDB size 140Gb. It seems to be working finally. @gheinrich how did you come up with the number of 500MB per thread per million keys? If i have **12GB** of memory on my Titan, that does not compute. Thanks mate, and thanks for the thread-tip!

**Comment by gheinrich** [comment]
Hi Tim, I was referring to CPU memory footprint for storing LMDB keys. The GPU memory is mostly a function of batch size and the number of parameters in your model and this doesn't depend on the number of data loader threads.

Today I have resumed working on this issue. I am trying to enable data loading from LMDB without storing keys at all, which would improve CPU memory consumption a lot. But I have troubles doing it in a multi-threaded environment. I'll keep you posted! Thanks for your patience!

**Comment by TimZaman** [comment]
But as @DmitryUlyanov noted with only 300k images and 256Gb CPU RAM and I have found out (by making it work with 1 thread), it does not seem to be related to the CPU's RAM.

**Comment by gheinrich** [comment]
The memory error fires before we load the model into GPU. There might be restrictions in place on @DmitryUlyanov 's system that limit the maximum amount of memory that a single process is able to use so the error might fire well before we hit the hard 256GB limit. By all means, if you find anything please let us know. Thanks in advance!

**Comment by gheinrich** [comment]
It turns out that Luajit limits heap memory to 2GB. So I am proposing a change to move LMDB keys from a Lua table to a `tds.Vec` (where items are allocated off the C heap and don't suffer from the Luajit 2GB heap limit).

I was able to train 1 epoch of Alexnet on 10 million 256x256 RGB images. It took 4h47m on a Titan X and GPU utilization was consistently over 95%.

@DmitryUlyanov @TimZaman can you try PR #686?

A further improvement would be to share LMDB keys between threads but I'd like to see if #686 is working for you first. Thanks!

**Comment by TimZaman** [comment]
1M images, 256x256, batch size 256, vanilla AlexNet, Titan X, 4 loader threads. and your pull request: **now works**. On Caffe it took around an hour per epoch. This seems to be taking 2 hours per epoch. But I have a  1200% luajit CPU usage now. GPU utilisation is still swinging a lot between 0 and 100 (that could be the 8tb hdd i am using).

**Comment by gheinrich** [comment]
Thank you very much for testing this Tim! Does the code look sensible to you? I wouldn't mind another pair of eyes to review this!

I've created #687 to see if we can tackle the perf issue.

**Comment by michaelholm-ce** [comment]
@gheinrich:  I'm experiencing this exact error.  I see you've closed this issue, but I'm still not sure what I can do to address the problem.  My dataset is a large lmdb set created by DIGITS.

I tried setting OMP_NUM_THREADS=1 and even setting batch_size=1.  The memory limits on the GPU and CPU do not grow very large, according to DIGITS.  

Will you please point me in the right direction here?  I can provide additional information, if needed.

**Comment by TimZaman** [comment]
@michaelholm-ce do you have **tds** installed as a lua package?

**Comment by michaelholm-ce** [comment]
@TimZaman:  I did not, but now do, and am trying it.  Thank you.
